### ISSUE TITLE
Default for includes to return errors may be a breaking change for app developers

### ISSUE THREAD

**Issue by aljachimiak** [oddnetworks/oddworks#233]
To handle a scenario (described in [oddworks Issue 160](https://github.com/oddnetworks/oddworks/issues/160)), [a new change to the handling](https://github.com/oddnetworks/oddworks/pull/229) of `?include` queries can cause a condition where an errors array is sent to the client.

### Issue:
This default will cause a breaking change in any deployments set to get the latest oddworks when they are deployed.

There is a way to change this default to send back partial includes that are found rather than the errors array.

## The Decision that needs to be made.

(1) Should the default be to send errors?  If this is the case, some deployments with 3rd party contractors developing apps, need to configures to send back partial includes.

*a different way to go about it:*

(2) Invert the default to always send partial includes.  Then the special configuration can be made to send back errors.  If this is the case, new deployments, (and current content-service) would need to configured to send back partial includes.

#### Please weigh in and:
- consider the work needed to make current deployments work
- think about the future of oddworks and the impact that will have on the library and it's deployments that haven't happened yet.
- and recommend what you think will be the proper course of action right now (and future actions if they  might be needed).

**Comment by pmcconnell** [comment]
If the path forward is errors then errors should be the default and no partial should be a part of that. Existing implementations should remain as they have been.

As I said when we discussed this, if I am writing a new app and I know there may be errors I will not use the include method and just handle it further downstream. If the apps are to handle these errors gracefully they will need to load any valid content in spite of errors. The way to do that is to skip the include and just do it manually. So if you need to write the code to work around errors anyway you may as well just make that code the golden path. If others adopt this method the include method is all but deprecated through lack of use.

Yes, partial returns would work around this but it is my understanding that the json spec says an error is an error and you can't do this. Either way I believe it would be a code smell to extract a response from an error.

**Comment by kixxauth** [comment]
The problem Al is addressing is that if we pull "includes may return errors" feature into Ooyala PV3 or ISC it would be a big problem, since, there are now 3 separate fleets of apps deployed on various instances of an oddworks server. Prior to this change going out, we need to inform all the devops teams to set the proper environment variable.

I'm actually going to walk back my opinion on this. It's probably not a big deal to get ahead of it and have all the appropriate devops teams set the environment variable before this goes out.

**Comment by brkattk** [comment]
If either way it requires a deploy, then adding an env var with it should be trivial. I say that not knowing how any of those are set up. As long as we communicate this properly we should be ok

**Comment by aljachimiak** [comment]
TL;DR;  leave the default to send errors if resources are missing.

Here's my take:
(1) This change is fixing an incorrect response.
(2) This incorrect response is being corrected at the beginning of a major release (4).
(3) To reduce cognitive load in the future, the default should be what we intend to be the default in the future.  I think this is to send errors.
(4) A way to be backwards compatible has been built in.  We should use this ability to fix the systems that absolutely *need* it.


>Prior to this change going out, we need to inform all the devops teams to set the proper environment variable.
I believe each of the deployed systems that will need the fix will need to have their codebase touched.  Rather than further burdening devops folks and managing environment errors it may be more efficient now (and in future deployments) to simply set the needed variable in a config.js.

**Comment by blainsmith** [comment]
Well are protected under semver since this fix falls into 4.x. For the currently deployed Ooyala instances they are on 3.x so their apps are fine as we publish. If they decide to upgrade then yes release notes need to be read by their devops team.

I think this should warrant finally starting a legit `CHANGELOG.md` file to start keeping track of the massive things going on.

As for the actual fix I think the default should be to send back errors. This not only makes most sense, but we, as the so-called experts, should make that call.

**Comment by aljachimiak** [comment]
>Ooyala instances they are on 3.x 

We'll have to look this over because there is more than just one instance.

**Comment by aljachimiak** [comment]
Consensus is to send errors as default
---------------------------------------------
